pkg/services/ngalert/migration/ualert.go

```diff
@@ -2,109 +2,120 @@ package migration
 }
 
-func (om *OrgMigration) migrateDashboard(ctx context.Context, dashID int64, alerts []*legacymodels.Alert) ([]models.AlertRule, error) {
-	info, err := om.migratedFolder(ctx, om.log, dashID)
+func (om *OrgMigration) migrateDashboard(ctx context.Context, dashID int64, alerts []*legacymodels.Alert) (*migmodels.DashboardUpgrade, error) {
+	dashboard, err := om.migrationStore.GetDashboard(ctx, om.orgID, dashID)
 	if err != nil {
-		return nil, fmt.Errorf("get or create migrated folder: %w", err)
+		if errors.Is(err, dashboards.ErrDashboardNotFound) {
+			// Historically, orphaned alerts are an exception to failOnError. We can revisit this if necessary.
+			om.log.Warn("Dashboard not found, alerts are orphaned", "dashboardId", dashID, "alertCount", len(alerts))
+			du := migmodels.NewDashboardUpgrade(dashID)
+			du.AddAlertErrors(dashboards.ErrDashboardNotFound, alerts...)
+			return du, nil
+		}
+		return nil, fmt.Errorf("get dashboard '%d': %w", dashID, err)
 	}
-	rules, err := om.migrateAlerts(ctx, alerts, *info)
+	l := om.log.New(
+		"dashboardTitle", dashboard.Title,
+		"dashboardUid", dashboard.UID,
+	)
+	l.Info("Migrating alerts for dashboard", "alertCount", len(alerts))
+
+	du := migmodels.NewDashboardUpgrade(dashID)
+	du.UID = dashboard.UID
+	du.Title = dashboard.Title
+	// nolint:staticcheck
+	du.FolderID = dashboard.FolderID
+
+	pairs, err := om.migrateAlerts(ctx, l, alerts, dashboard)
 	if err != nil {
-		return nil, fmt.Errorf("migrate and save alerts: %w", err)
+		if om.failOnError {
+			return nil, err
+		}
+		l.Warn("Failed to migrate dashboard alerts", "alertCount", len(alerts), "error", err)
+		du.AddAlertErrors(err, alerts...)
+		return du, nil
 	}
 
-	return rules, nil
+	for _, pair := range pairs {
+		du.MigratedAlerts[pair.LegacyRule.PanelID] = pair
+	}
+	return du, nil
 }
 
-func (om *OrgMigration) migrateOrgAlerts(ctx context.Context) error {
+func (om *OrgMigration) migrateOrgAlerts(ctx context.Context) ([]*migmodels.DashboardUpgrade, error) {
 	mappedAlerts, cnt, err := om.migrationStore.GetOrgDashboardAlerts(ctx, om.orgID)
 	if err != nil {
-		return fmt.Errorf("load alerts: %w", err)
+		return nil, fmt.Errorf("load alerts: %w", err)
 	}
 	om.log.Info("Alerts found to migrate", "alerts", cnt)
 
+	dashboardUpgrades := make([]*migmodels.DashboardUpgrade, 0, len(mappedAlerts))
 	for dashID, alerts := range mappedAlerts {
-		rules, err := om.migrateDashboard(ctx, dashID, alerts)
-		if err != nil {
-			return fmt.Errorf("migrate and save dashboard '%d': %w", dashID, err)
-		}
-
-		if len(rules) > 0 {
-			om.log.Debug("Inserting migrated alert rules", "count", len(rules))
-			err := om.migrationStore.InsertAlertRules(ctx, rules...)
-			if err != nil {
-				return fmt.Errorf("insert alert rules: %w", err)
-			}
+		du, err := om.migrateDashboard(ctx, dashID, alerts)
+		if err != nil && om.failOnError {
+			return nil, fmt.Errorf("migrate dashboard '%d': %w", dashID, err)
 		}
+		dashboardUpgrades = append(dashboardUpgrades, du)
```

Comment:
This is counterintuitive because err != nil path is not always terminated. THere needs to be either a comment explaining a situation when we would like to add `du` to the result when `err != nil` or error handling piece rewritten to terminate loop in case of error.

Reply:
Yah, this is not right. I ended up refactoring towards what we discussed: removing failOnError, collecting the errors and failing at the end if any errors exist.

https://github.com/grafana/grafana/pull/79719/files#diff-0f24ee88a6a53f55b03ad33be23aea4d8674537474027be555073431d6c8d34bR181-R198